### PR TITLE
Upgrade packer and send more FF data 

### LINKED ISSUE #197: Upgrade packer to new version 2 to allow MCU to send more FF data on UART Push link
We currently send an average value of optical power from all channels of each FF device. With a new auto-encoder, we will send optical power per channel. 

Additionally, Rui requested us to send outputs from transmit output disable register as well. 

**Describe the solution you'd like**
Follow a new auto-encoder as in [here](https://gitlab.com/apollo-lhc/FW/SM_ZYNQ_FW/-/blob/develop/src/CM_interface/CM_Monitoring/data_format.md?ref_type=heads) and modify [this function](https://github.com/apollo-lhc/cm_mcu/blob/master/projects/cm_mcu/ZynqMonTask.c#L54). 

Then, just add more zynqmon data steams as we have done before.

### PR DESCRIPTION
**Pull request by pkotamnives**
This PR will close #197 

The v2 packer has been implemented, and it is compatible with the current set of data steams. This has been checked on Apollo205, and BUTool `status` looks consistent with what has been shown before. 

Next step is to send all FF-ch optical power outputs, `ff xmit on` output, and more ff register outputs. Some have also been implemented via CLI. The following list is what missing :
- peak temperature? currently we send FF temperature per device 
- per rx ch of Received Optical Power(ROP) (currently we send average value of ROP per device)
- LOL has an implementation via CLI `ff_cdr_lol`
- LOS has an implementation via CLI `ff_los_alarm`
- missing laser fault (0x9-0x10 for 12-ch or 0x4 for 4-ch)

### CODE REVIEW

**Comment by pwittich** [comment]
@pkotamnives how did you test this? I thought this required a change in the Zynq FW

**Comment by pkotamnives** [comment]
> @pkotamnives how did you test this? I thought this required a change in the Zynq FW

Hi Peter, I first checked the BUTool `status` from cm_mcu.bin that is up-to-date with the master. Then, I reprogrammed the mcu firmware with the cm_mcu.bin after changing the packer to v2 (no change to data steams) and I see the same `status` unchanged. 

there was a bug that I just fixed though. I can re-check again.

**Comment by pwittich** [comment]
> > @pkotamnives how did you test this? I thought this required a change in the Zynq FW
> 
> Hi Peter, I first checked the BUTool `status` from cm_mcu.bin that is up-to-date with the master. Then, I reprogrammed the mcu firmware with the cm_mcu.bin after changing the packer to v2 (no change to data steams) and I see the same `status` unchanged.
> 
> there was a bug that I just fixed though. I can re-check again.

yes please, I think we need a new firmware from @dgastler for this to work. Is it possible you were looking at stale values?

**Comment by pkotamnives** [comment]
Hi @pwittich and @dgastler, I have reprogrammed mcu with the master cm_mcu.bin. The BUTool `status` looks good. Then I changed it to the cm_mcu.bin from the latest commit in this branch. I found that the MCU_UPTIME and other fields were frozen after the packer is changed.
This seems to explain why I thought the code works but actually it's not. Could you update the Zynq FW for this purpose?

File: projects/cm_mcu/ZynqMonTask.c [hunk 1]
```diff
@@ -528,9 +548,9 @@ void zm_send_data(struct zynqmon_data_t data[])
 {
   // https://docs.google.com/spreadsheets/d/1E-JD7sRUnkbXNqfgCUgriTZWfCXark6IN9ir_9b362M/edit#gid=0
   for (int i = 0; i < ZMON_VALID_ENTRIES; ++i) {
```
Comment (pwittich) [inline projects/cm_mcu/ZynqMonTask.c:550]: this link is obsolete I think

File: projects/cm_mcu/ZynqMonTask.c [hunk 2]
```diff
@@ -528,9 +548,9 @@ void zm_send_data(struct zynqmon_data_t data[])
...
-    uint8_t message[4];
+    uint8_t message[5];
```
Comment (pwittich) [inline projects/cm_mcu/ZynqMonTask.c:551]: maybe introduce some #define constant for the size of the message

File: projects/cm_mcu/ZynqMonTask.c [hunk 3]
```diff
@@ -43,27 +43,47 @@
...
+#define SENSOR_MESSAGE_START_OF_FRAME_NIB    2
+#define SENSOR_MESSAGE_START_OF_FRAME_NIB_V2 3
+#define RESERVED_DATA_V2                     0x9 // 0b1001
+#define SENSOR_MESSAGE_DATA_FRAME_NIB        0
+#define SENSOR_MESSAGE_HEADER_OFFSET         6
+#define SENSOR_SIX_BITS                      0x3F
+#define DATA_FOUR_BITS                       0xF
+#define DATA_TWO_BITS_MASK0                  0x3  // a mask for 0b11
+#define DATA_TWO_BITS_MASK1                  0xC  // a mask for 0b1100
+#define DATA_TWO_BITS_MASK2                  0x30 // a mask for 0b110000
```
[the remark this answers is not in the document]
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c:55]: Some old masks relevant to the code are kept but the rest is removed. The V2 tags are removed from certain parts. Done.

File: projects/cm_mcu/ZynqMonTask.c (hunk 1 above)
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c:550]: Removed. 
Done.

File: projects/cm_mcu/ZynqMonTask.c (hunk 2 above)
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c:551]: Sounds good, will do. 
Done.

File: projects/cm_mcu/ZynqMonTask.c (hunk 1 above)
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c:550]: Though, do you think that we should make a v2 spreadsheet after this PR is done?

Comment (pwittich) [reply projects/cm_mcu/ZynqMonTask.c:550]: we should consider having some automated way of doing this, though I think it is not trivial

**Comment by pwittich** [comment]
Thanks for making the changes! I think there might still be some issues; look at the grafana (I fixed the setup so that it shows up again):
https://www.classe.cornell.edu/cms/grafana/psbATCA/d/YVW4b-n4k/rev2-cm-dashboard?orgId=2&refresh=30s&from=1704213966757&to=1704215766757&var-WhichApollo=AP205
the fact that the values glitch to zero (or non-zero) suggests that we still have some issues with overwriting memory to me....

**Comment by pkotamnives** [comment]
**Update on rollover/overwrite issue:** We found that it is a bug in Task.h (i.e. ZM_NUM_ENTRIES = 256 ) that prevents higher address than 0xFF to be written and the program also crashed when more entries than 256 are sent or overwrote to the first table otherwise. We also changed the sensor type from uint8_t to uint16_t just to be consistent with the data type. 

Other [bugs and corrections](https://github.com/apollo-lhc/cm_mcu/issues/204) as well as the addition of [ff_laser_fault ](https://github.com/apollo-lhc/cm_mcu/tree/dev_CLI_laser_fault)to a list of data sent to Zynq will be merged into this branch from https://github.com/apollo-lhc/cm_mcu/tree/test_packer2rollover

File: sm_cm_config/src/xml_generate.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python
+#!/ usr / bin / env python
```
Comment (pwittich) [inline sm_cm_config/src/xml_generate.py:1]: this is a bad change, please revert

File: projects/cm_mcu/LocalTasks.c [hunk 5]
```diff
@@ -173,10 +173,10 @@ struct sm_command_t sm_command_ffldaq_f1[] = {
     {1, 0x00, 0x16, 2, "FF_TEMPERATURE", 0xff, "C", PM_STATUS},
     {1, 0x00, 0x03, 1, "FF_LOS_ALARM", 0xff, "", PM_STATUS},
     {1, 0x00, 0x05, 1, "FF_CDR_LOL_ALARM", 0xff, "", PM_STATUS},
-    {2, 0x00, 0x22, 1, "FF_CH01_OPT_POW", 0xff, "mw", PM_STATUS}, // read 4 Rx-ch registers with increasing addresses
-    {2, 0x00, 0x24, 1, "FF_CH02_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x00, 0x26, 1, "FF_CH03_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x00, 0x28, 1, "FF_CH04_OPT_POW", 0xff, "mw", PM_STATUS},
+    {2, 0x00, 0x22, 2, "FF_CH01_OPT_POW", 0xff, "mw", PM_STATUS}, // read 4 Rx-ch registers with increasing addresses
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:176]: mw -> mW (milli-Watt) four times

File: projects/cm_mcu/LocalTasks.c [hunk 6]
```diff
@@ -230,18 +230,18 @@ struct sm_command_t sm_command_fflot_f1[] = {
...
-    {2, 0x01, 0xde, 1, "FF_CH04_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xdc, 1, "FF_CH05_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xda, 1, "FF_CH06_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xd8, 1, "FF_CH07_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xd6, 1, "FF_CH08_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xd4, 1, "FF_CH09_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xd2, 1, "FF_CH10_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xd0, 1, "FF_CH11_OPT_POW", 0xff, "mw", PM_STATUS},
-    {2, 0x01, 0xce, 1, "FF_CH12_OPT_POW", 0xff, "mw", PM_STATUS},
+    {2, 0x01, 0xe4, 2, "FF_CH01_OPT_POW", 0xff, "mw", PM_STATUS}, // read 12 Rx-ch registers  with decreasing addresses
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:233]: again mw-> mW

File: projects/cm_mcu/LocalTasks.c [hunk 7]
```diff
@@ -374,18 +374,18 @@ struct sm_command_t sm_command_fflot_f2[] = {
...
```
Comment (pwittich) [inline projects/cm_mcu/LocalTasks.c:377]: mw -> mW

File: projects/cm_mcu/ZynqMonTask.c [hunk 8]
```diff
@@ -43,27 +44,37 @@
 // Register List
 // See Google Docs, 'CM uC Sensor register map'
 
-#define SENSOR_MESSAGE_START_OF_FRAME_NIB 2
-#define SENSOR_MESSAGE_DATA_FRAME_NIB     0
-#define SENSOR_MESSAGE_HEADER_OFFSET      6
-#define SENSOR_SIX_BITS                   0x3F
+#define SENSOR_MESSAGE_START_OF_FRAME_NIB    2
+#define SENSOR_MESSAGE_START_OF_FRAME_NIB_V2 3
+#define RESERVED_DATA                        0x9 // 0b1001
```
Comment (pwittich) [inline projects/cm_mcu/ZynqMonTask.c:49]: fwiw, this just a dummy value. We don't need to put this in

File: projects/cm_mcu/ZynqMonTask.c [hunk 9]
```diff
@@ -528,9 +582,9 @@ void zm_send_data(struct zynqmon_data_t data[])
 {
   // https://docs.google.com/spreadsheets/d/1E-JD7sRUnkbXNqfgCUgriTZWfCXark6IN9ir_9b362M/edit#gid=0
   for (int i = 0; i < ZMON_VALID_ENTRIES; ++i) {
-    uint8_t message[4];
+    uint8_t message[5];
     format_data(data[i].sensor, data[i].data.us, message);
-    for (int j = 0; j < 4; ++j) {
+    for (int j = 0; j < 5; ++j) {
```
Comment (pwittich) [inline projects/cm_mcu/ZynqMonTask.c:587]: add something for magic number 5 in various places

**Review by pwittich — CHANGES_REQUESTED** [review]
take a look at these changes

File: sm_cm_config/data/PL_MEM_CM_rev2.yml [hunk 10]
```diff
@@ -37,8 +37,9 @@ config:
     postfixes: 
       - Temp_C
       - OptPow
```
[the remark this answers is not in the document]
Comment (pkotamnives) [reply sm_cm_config/data/PL_MEM_CM_rev2.yml:39]: Hmm I think all of them are uint16_t as in https://github.com/apollo-lhc/cm_mcu/blob/a5c8a0456e0b7c06ab7fa46dafcad4a75b40f83a/projects/cm_mcu/Tasks.h#L201-L203. I found that the OptPow is a two-byte value with a unit in 0.1 microwatts

[the remark this answers is not in the document]
Comment (pwittich) [reply sm_cm_config/data/PL_MEM_CM_rev2.yml:39]: apparently this is incorrect! good we don't have to deal with it immediately

File: projects/cm_mcu/LocalTasks.c (hunk 5 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c]: it's actually in the unit of 0.1uW

File: projects/cm_mcu/ZynqMonTask.c (hunk 8 above)
Comment (pkotamnives) [reply projects/cm_mcu/ZynqMonTask.c:49]: do you mean typing them directly to the code?

**Review by pwittich — APPROVED** [review]

**Pull request merged**